# Car Price Prediction Using Machine Learning

## Introduction

In this project, we will analyze car data and build a machine learning
model to predict the selling price of a used car.

We will first understand and clean the data, then perform some data
analysis and visualization.

After that, we will train regression models and compare their results.

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

## Loading the Dataset

In this step, we will load the car dataset using pandas and store it in a
DataFrame for further analysis.

In [3]:
df=pd.read_csv("car_data.csv")

In [4]:
df.head()

,Car_Name,Year,Selling_Price,Present_Price,Kms_Driven,Fuel_Type,Seller_Type,Transmission,Owner
0,ritz,2014,3.35,5.59,27000,Petrol,Dealer,Manual,0
1,sx4,2013,4.75,9.54,43000,Diesel,Dealer,Manual,0
2,ciaz,2017,7.25,9.85,6900,Petrol,Dealer,Manual,0
3,wagon r,2011,2.85,4.15,5200,Petrol,Dealer,Manual,0
4,swift,2014,4.60,6.87,42450,Diesel,Dealer,Manual,0


## Checking the Dataset Size

Here, we will check the number of rows and columns in the dataset.

In [5]:
df.shape

(301, 9)

## Checking the Column Names

In this step, we will check the names of all the columns available in
the dataset.

In [6]:
df.columns

Index(['Car_Name', 'Year', 'Selling_Price', 'Present_Price', 'Kms_Driven',
       'Fuel_Type', 'Seller_Type', 'Transmission', 'Owner'],
      dtype='object')

## Checking Dataset Information

Here, we will check the data types, number of non-null values, and
overall information about the dataset.

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 301 entries, 0 to 300
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Car_Name       301 non-null    object 
 1   Year           301 non-null    int64  
 2   Selling_Price  301 non-null    float64
 3   Present_Price  301 non-null    float64
 4   Kms_Driven     301 non-null    int64  
 5   Fuel_Type      301 non-null    object 
 6   Seller_Type    301 non-null    object 
 7   Transmission   301 non-null    object 
 8   Owner          301 non-null    int64  
dtypes: float64(2), int64(3), object(4)
memory usage: 21.3+ KB


## Checking Data Types

We will check the data type of each column to understand which columns
contain numerical data and which contain categorical data.

In [8]:
df.dtypes

Car_Name          object
Year               int64
Selling_Price    float64
Present_Price    float64
Kms_Driven         int64
Fuel_Type         object
Seller_Type       object
Transmission      object
Owner              int64
dtype: object

## Checking Missing Values

In this step, we will check whether the dataset contains any missing
values in the columns.

In [9]:
df.isnull().sum()

Car_Name         0
Year             0
Selling_Price    0
Present_Price    0
Kms_Driven       0
Fuel_Type        0
Seller_Type      0
Transmission     0
Owner            0
dtype: int64

## Statistical Summary

Here, we will look at the basic statistical summary of the numerical
columns, such as the mean, minimum, maximum, and quartile values.

In [10]:
df.describe()

,Year,Selling_Price,Present_Price,Kms_Driven,Owner
count,301.000000,301.000000,301.000000,301.000000,301.000000
mean,2013.627907,4.661296,7.628472,36947.205980,0.043189
std,2.891554,5.082812,8.644115,38886.883882,0.247915
min,2003.000000,0.100000,0.320000,500.000000,0.000000
25%,2012.000000,0.900000,1.200000,15000.000000,0.000000
50%,2014.000000,3.600000,6.400000,32000.000000,0.000000
75%,2016.000000,6.000000,9.900000,48767.000000,0.000000
max,2018.000000,35.000000,92.600000,500000.000000,3.000000


## Initial Data Understanding

The dataset contains 301 rows and 9 columns.

There are no missing values in the dataset.

The dataset contains both numerical and categorical columns.

The numerical columns are Year, Selling_Price, Present_Price, Kms_Driven,
and Owner.

The categorical columns are Car_Name, Fuel_Type, Seller_Type, and
Transmission.

Selling_Price will be used as the target variable for predicting the
price of a used car.

## Exploring Categorical Columns

In this step, we will check the unique values and value counts of the
categorical columns in the dataset.

In [11]:
categorical_columns = ["Fuel_Type", "Seller_Type", "Transmission", "Owner"]

for column in categorical_columns:
    print("Column:", column)
    print("Unique values:", df[column].unique())
    print("Value counts:")
    print(df[column].value_counts())
    print("-" * 40)

Column: Fuel_Type
Unique values: ['Petrol' 'Diesel' 'CNG']
Value counts:
Fuel_Type
Petrol    239
Diesel     60
CNG         2
Name: count, dtype: int64
----------------------------------------
Column: Seller_Type
Unique values: ['Dealer' 'Individual']
Value counts:
Seller_Type
Dealer        195
Individual    106
Name: count, dtype: int64
----------------------------------------
Column: Transmission
Unique values: ['Manual' 'Automatic']
Value counts:
Transmission
Manual       261
Automatic     40
Name: count, dtype: int64
----------------------------------------
Column: Owner
Unique values: [0 1 3]
Value counts:
Owner
0    290
1     10
3      1
Name: count, dtype: int64
----------------------------------------


## Checking Duplicate Rows

Here, we will check whether the dataset contains any duplicate rows.
Duplicate rows can affect the result of a machine learning model.

In [14]:
df.duplicated().sum()

np.int64(2)

## Duplicate Rows Result

The duplicate row count is checked to understand whether any repeated
records are present in the dataset.

In [15]:
df = df.drop_duplicates()

print("Dataset shape after removing duplicates:", df.shape)

Dataset shape after removing duplicates: (299, 9)


## Exploring Numerical Columns

In this step, we will check the minimum, maximum, average, and other
statistical values of the numerical columns.

In [16]:
numerical_columns = [
    "Year",
    "Selling_Price",
    "Present_Price",
    "Kms_Driven",
    "Owner"
]

df[numerical_columns].describe()

,Year,Selling_Price,Present_Price,Kms_Driven,Owner
count,299.000000,299.000000,299.000000,299.000000,299.000000
mean,2013.615385,4.589632,7.541037,36916.752508,0.043478
std,2.896868,4.984240,8.567887,39015.170352,0.248720
min,2003.000000,0.100000,0.320000,500.000000,0.000000
25%,2012.000000,0.850000,1.200000,15000.000000,0.000000
50%,2014.000000,3.510000,6.100000,32000.000000,0.000000
75%,2016.000000,6.000000,9.840000,48883.500000,0.000000
max,2018.000000,35.000000,92.600000,500000.000000,3.000000


## Creating Car Age

The dataset contains the manufacturing year of each car.

We will create a new column called Car_Age by subtracting the car year
from the current year.

In [17]:
current_year = 2026

df["Car_Age"] = current_year - df["Year"]

df[["Year", "Car_Age"]].head()

C:\Users\Arun\AppData\Local\Temp\ipykernel_13444\3384860266.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["Car_Age"] = current_year - df["Year"]


,Year,Car_Age
0,2014,12
1,2013,13
2,2017,9
3,2011,15
4,2014,12


## Checking Car Age Values

Here, we will check the basic statistics of the newly created Car_Age
column.

In [18]:
df["Car_Age"].describe()

count    299.000000
mean      12.384615
std        2.896868
min        8.000000
25%       10.000000
50%       12.000000
75%       14.000000
max       23.000000
Name: Car_Age, dtype: float64